app/src/main/java/com/bootcamp/concrete/magicdeck/ui/activity/CatalogActivity.kt

```diff
@@ -0,0 +1,98 @@
+import androidx.lifecycle.observe
+import androidx.recyclerview.widget.GridLayoutManager
+import com.bootcamp.concrete.magicdeck.R
+import com.bootcamp.concrete.magicdeck.data.domain.Card
+import com.bootcamp.concrete.magicdeck.data.domain.CardListItem
+import com.bootcamp.concrete.magicdeck.ui.adapter.CardsListAdapter
+import com.bootcamp.concrete.magicdeck.ui.adapter.DividerItemDecoration
+import com.bootcamp.concrete.magicdeck.viewmodel.CatalogViewModel
+import com.bootcamp.concrete.magicdeck.viewmodel.CatalogViewModelFactory
+import com.bootcamp.concrete.magicdeck.viewmodel.CatalogViewModelState
+import kotlinx.android.synthetic.main.activity_catalog.cards_catalog
+
+class CatalogActivity : AppCompatActivity() {
+
+    private val cards = ArrayList<CardListItem>()
+    private val catalogViewModel: CatalogViewModel by viewModels {
+        CatalogViewModelFactory()
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_catalog)
+        setUpList()
+        catalogViewModel.getInitialCards()
+        catalogViewModel.getViewState().observe(this) {
+            when (it) {
+                is CatalogViewModelState.NavigateToCarousel -> startCardCarouselActivity(it.card)
+                is CatalogViewModelState.ListCards -> listCards(it.cardListItems)
+                is CatalogViewModelState.Failure -> showErrorMessage()
+                is CatalogViewModelState.Error -> showErrorMessage(it.stringId)
+            }
+        }
+
+
+    }
+
+    private fun setUpList() {
+        cards_catalog.addItemDecoration(
+            DividerItemDecoration(
+                resources.getInteger(R.integer.list_divider_size)
+            )
+        )
+        cards_catalog.adapter = CardsListAdapter(cards, this@CatalogActivity) {
+            startCardCarouselActivity(it)
+        }
+        val layoutManager =
+            GridLayoutManager(this@CatalogActivity, resources.getInteger(R.integer.list_span_size))
+        changeListItemSpanSize(layoutManager)
+        cards_catalog.layoutManager = layoutManager
+    }
+
+    private fun changeListItemSpanSize(layoutManager: GridLayoutManager) {
+        layoutManager.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
+            override fun getSpanSize(position: Int): Int {
+                cards_catalog.adapter?.let {
+                    return it.getItemViewType(position)
+                }
+                return 1
+            }
+        }
+    }
+
+    private fun startCardCarouselActivity(card: Card) {
+        val intent = Intent(this@CatalogActivity, CardCarouselActivity::class.java)
+        intent.putExtra("card", card)
+        startActivity(intent)
+    }
+
+    private fun listCards(list: List<CardListItem>) {
+        cards.addAll(list)
+        cards_catalog.adapter?.notifyItemRangeInserted(
+            cards.size - list.size,
+            list.size
+        )
```

Comment:
Se for null? não tem resposta nenhum pra o usuário?
`?: run {  }`

Reply:
essa é uma dúvia que tenho tuiliznado o koltin.... isso de null safety as vezes fica esquisito, pq esse método é chamado depois da criação e configuração da lista, então existe um adapter. Mas o kotlin não eh capaz de vê isso, por isso coloquei isso, não sei o que seria legal de mostrar ao usuário caso a lista não tenha um adapter nesse ponto.

Criar o adapter? mas se eu colocar como elvis notation ele vai adicionar as cartas de toda forma.